File: himsha-sdk/src/pubkey.ts
import { createHash } from 'crypto';

const BASE58_ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

function base58Encode(bytes: Uint8Array): string {
  let num = BigInt('0x' + Buffer.from(bytes).toString('hex') || '0');
  let result = '';
  const base = BigInt(58);
  while (num > 0n) {
    result = BASE58_ALPHABET[Number(num % base)] + result;
    num = num / base;
  }
  for (const byte of bytes) {
    if (byte === 0) result = '1' + result;
    else break;
  }
  return result;
}

function base58Decode(str: string): Uint8Array {
  let num = 0n;
  const base = BigInt(58);
  for (const char of str) {
    const idx = BASE58_ALPHABET.indexOf(char);
    if (idx < 0) throw new Error(`Invalid base58 character: ${char}`);
    num = num * base + BigInt(idx);
  }
  const bytes: number[] = [];
  while (num > 0n) {
    bytes.unshift(Number(num & 0xffn));
    num >>= 8n;
  }
  for (const char of str) {
    if (char === '1') bytes.unshift(0);
    else break;
  }
  return new Uint8Array(bytes);
}

/** 32-byte public key. */
export class HimshaPublicKey {
  private readonly _bytes: Uint8Array;

  constructor(bytes: Uint8Array | string) {
    if (typeof bytes === 'string') {
      this._bytes = base58Decode(bytes);
    } else {
      this._bytes = bytes;
    }
    if (this._bytes.length !== 32) {
      throw new Error(`PublicKey must be 32 bytes, got ${this._bytes.length}`);
    }
  }

  static fromBase58(s: string): HimshaPublicKey {
    return new HimshaPublicKey(s);
  }

  static fromSeed(seed: string): HimshaPublicKey {
    const hash = createHash('sha256').update(seed, 'utf8').digest();
    return new HimshaPublicKey(new Uint8Array(hash));
  }

  static findProgramAddress(seeds: Uint8Array[], programId: HimshaPublicKey): [HimshaPublicKey, number] {
    for (let nonce = 255; nonce >= 0; nonce--) {
      const hasher = createHash('sha256');
      for (const seed of seeds) hasher.update(seed);
      hasher.update(new Uint8Array([nonce]));
      hasher.update(programId.toBytes());
      hasher.update(Buffer.from('himsha::pda', 'utf8'));
      return [new HimshaPublicKey(new Uint8Array(hasher.digest())), nonce];
    }
    throw new Error('Could not find program address');
  }

  toBase58(): string { return base58Encode(this._bytes); }
  toBytes(): Uint8Array { return this._bytes; }
  toString(): string { return this.toBase58(); }

  equals(other: HimshaPublicKey): boolean {
    return this._bytes.every((b, i) => b === other._bytes[i]);
  }
}

/** Well-known built-in program IDs. */
export const PROGRAM_IDS = {
  system:      HimshaPublicKey.fromSeed('himsha::system_program'),
  token:       HimshaPublicKey.fromSeed('himsha::token_program'),
  ata:         HimshaPublicKey.fromSeed('himsha::ata_program'),
  swap:        HimshaPublicKey.fromSeed('himsha::swap_program'),
  lending:     HimshaPublicKey.fromSeed('himsha::lending_program'),
  nftMetadata: HimshaPublicKey.fromSeed('himsha::nft_metadata_program'),
  runes:       HimshaPublicKey.fromSeed('himsha::runes_program'),
  moneyMarket: HimshaPublicKey.fromSeed('himsha::money_market_program'),
  vault:       HimshaPublicKey.fromSeed('himsha::vault_program'),
  oracle:      HimshaPublicKey.fromSeed('himsha::oracle_program'),
} as const;

File: himsha-sdk-dart/lib/src/pubkey.dart
import 'dart:typed_data';
import 'package:characters/characters.dart';
import 'package:crypto/crypto.dart';

const String _alphabet = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

String _base58Encode(Uint8List bytes) {
  BigInt num = BigInt.zero;
  for (final b in bytes) {
    num = num * BigInt.from(256) + BigInt.from(b);
  }
  var result = '';
  final base = BigInt.from(58);
  while (num > BigInt.zero) {
    final rem = num.remainder(base).toInt();
    result = _alphabet[rem] + result;
    num = num ~/ base;
  }
  for (final b in bytes) {
    if (b == 0) {
      result = '1$result';
    } else {
      break;
    }
  }
  return result;
}

Uint8List _base58Decode(String s) {
  BigInt num = BigInt.zero;
  final base = BigInt.from(58);
  for (final c in s.characters) {
    final idx = _alphabet.indexOf(c);
    if (idx < 0) throw ArgumentError('Invalid base58 character: $c');
    num = num * base + BigInt.from(idx);
  }
  final bytes = <int>[];
  while (num > BigInt.zero) {
    bytes.insert(0, (num & BigInt.from(0xff)).toInt());
    num >>= 8;
  }
  for (final c in s.characters) {
    if (c == '1') {
      bytes.insert(0, 0);
    } else {
      break;
    }
  }
  return Uint8List.fromList(bytes);
}

/// 32-byte identifier for accounts and programs on HIMSHA Network.
class HimshaPublicKey {
  final Uint8List bytes;

  HimshaPublicKey(this.bytes) {
    if (bytes.length != 32) {
      throw ArgumentError('PublicKey must be 32 bytes, got ${bytes.length}');
    }
  }

  /// Create from base58 string.
  factory HimshaPublicKey.fromBase58(String s) => HimshaPublicKey(_base58Decode(s));

  /// Deterministic key from a UTF-8 seed string.
  factory HimshaPublicKey.fromSeed(String seed) {
    final hash = sha256.convert(seed.codeUnits).bytes;
    return HimshaPublicKey(Uint8List.fromList(hash));
  }

  /// Derive a Program Derived Address.
  static (HimshaPublicKey, int) findProgramAddress(
    List<Uint8List> seeds,
    HimshaPublicKey programId,
  ) {
    for (int nonce = 255; nonce >= 0; nonce--) {
      final digest = sha256.convert([
        for (final s in seeds) ...s,
        nonce,
        ...programId.bytes,
        ...'himsha::pda'.codeUnits,
      ]);
      return (HimshaPublicKey(Uint8List.fromList(digest.bytes)), nonce);
    }
    throw StateError('Could not find program address');
  }

  String toBase58() => _base58Encode(bytes);

  @override
  String toString() => toBase58();

  @override
  bool operator ==(Object other) =>
      other is HimshaPublicKey &&
      bytes.length == other.bytes.length &&
      List.generate(32, (i) => bytes[i] == other.bytes[i]).every((b) => b);

  @override
  int get hashCode => Object.hashAll(bytes);
}

/// Well-known program IDs (deterministic).
class ProgramIds {
  static final system      = HimshaPublicKey.fromSeed('himsha::system_program');
  static final token       = HimshaPublicKey.fromSeed('himsha::token_program');
  static final ata         = HimshaPublicKey.fromSeed('himsha::ata_program');
  static final swap        = HimshaPublicKey.fromSeed('himsha::swap_program');
  static final lending     = HimshaPublicKey.fromSeed('himsha::lending_program');
  static final nftMetadata = HimshaPublicKey.fromSeed('himsha::nft_metadata_program');
}
